ui/src/treegrid.tsx

```diff
@@ -0,0 +1,247 @@
+import ErrorIcon from "@mui/icons-material/Error";
+import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
+import KeyboardArrowRightIcon from "@mui/icons-material/KeyboardArrowRight";
+import CircularProgress from "@mui/material/CircularProgress";
+import IconButton from "@mui/material/IconButton";
+import Typography from "@mui/material/Typography";
+import { ReactNode, useCallback } from "react";
+import { useImmer } from "use-immer";
+
+export type TreeGridId = string | number;
+
+export type TreeGridRowData = {
+  [key: string]: string | number | JSX.Element;
+};
+
+export type TreeGridItem = {
+  children?: TreeGridId[];
+  data: TreeGridRowData;
+};
+
+export type TreeGridData = {
+  [key: TreeGridId]: TreeGridItem;
+};
+
+export type TreeGridColumnData = {
+  key: string;
+  width: string | number;
+  title?: string | JSX.Element;
+};
+
+export type TreeGridProps = {
+  columns: TreeGridColumnData[];
+  data: TreeGridData;
+  prefixElements?: (id: TreeGridId, data: TreeGridRowData) => ReactNode;
+  loadChildren?: (id: TreeGridId) => Promise<string>;
+};
+
+export function TreeGrid(props: TreeGridProps) {
+  const [state, updateState] = useImmer<TreeGridState>(
+    new Map<TreeGridId, TreeGridItemState>()
+  );
+
+  const toggleExpanded = useCallback((id: TreeGridId) => {
+    updateState((draft) => {
+      const itemState = draft.get(id) || {
+        status: Status.Collapsed,
+      };
+      switch (itemState.status) {
+        case Status.Loading:
+          return;
+        case Status.Expanded:
+          itemState.status = Status.Collapsed;
+          break;
+        default:
+          if (!props.loadChildren || !!itemState.loaded) {
+            itemState.status = Status.Expanded;
+          } else {
+            itemState.status = Status.Loading;
+            props
+              .loadChildren(id)
+              .then(() => {
+                updateState((draft) => {
+                  draft.set(id, {
+                    status: Status.Expanded,
+                    loaded: true,
+                  });
+                });
+              })
+              .catch((error) => {
+                updateState((draft) => {
+                  draft.set(id, {
+                    status: Status.Failed,
+                    errorMessage: error,
+                  });
+                });
+              });
+          }
+      }
+      draft.set(id, itemState);
+    });
+  }, []);
```

Comment:
Should this be invalidated by `props.loadChildren` changing?

Reply:
Done.